presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp

```diff
@@ -234,6 +258,160 @@ TEST_P(HttpTestSuite, timedOutRequests) {
+
+  HttpClientFactory clientFactory;
+  auto client1 = clientFactory.newClient(
+      serverAddress,
+      std::chrono::milliseconds(5'000),
+      std::chrono::milliseconds(0),
+      useHttps,
+      memoryPool);
+  auto client2 = clientFactory.newClient(
+      serverAddress,
+      std::chrono::milliseconds(2'950),
+      std::chrono::milliseconds(0),
+      useHttps,
+      memoryPool);
+  auto client3 = clientFactory.newClient(
+      serverAddress,
+      std::chrono::milliseconds(900),
+      std::chrono::milliseconds(0),
+      useHttps,
+      memoryPool);
+  auto client4 = clientFactory.newClient(
+      serverAddress,
+      std::chrono::milliseconds(7'000),
+      std::chrono::milliseconds(2'000),
+      useHttps,
+      memoryPool);
+  request1->customFunc = []() {
+    // Holds the thread for 3s to fully occupy the server io pool (size 1), so
+    // that the second request shall wait in io pool queue for at least 3s
+    std::this_thread::sleep_for(std::chrono::milliseconds(3'000));
+  };
+
+  request1->maxWaitMillis = 10;
+  request2->maxWaitMillis = 10;
+  request3->maxWaitMillis = 1000;
+  request4->maxWaitMillis = 10;
+
+  // TODO: DBG REMOVE
+  request1->name = "request1";
+  request2->name = "request2";
+  request3->name = "request3";
+  request4->name = "request4";
+
+  auto promisePair1 = folly::makePromiseContract<bool>();
+  request1->requestPromise = std::move(promisePair1.first);
+  auto promisePair2 = folly::makePromiseContract<bool>();
+  request2->requestPromise = std::move(promisePair2.first);
+  auto promisePair3 = folly::makePromiseContract<bool>();
+  request3->requestPromise = std::move(promisePair3.first);
+  auto promisePair4 = folly::makePromiseContract<bool>();
+  request4->requestPromise = std::move(promisePair4.first);
+
+  std::thread thread1([&]() {
+    // Time 00:00.000
+    auto responseFuture = sendGet(client1.get(), "/async/msg1");
+    try {
+      std::move(promisePair1.second).wait();
+      auto response = std::move(responseFuture).get();
+      ASSERT_EQ(response->headers()->getStatusCode(), http::kHttpOk);
+      ASSERT_EQ(bodyAsString(*response, memoryPool.get()), "Timedout");
+      ASSERT_EQ(request1->requestStatus, kStatusValid);
+    } catch (...) {
+      FAIL();
+    }
+  });
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  std::thread thread2([&]() {
+    // Time 00:00.100
+    auto responseFuture = sendGet(client2.get(), "/async/msg2");
+    try {
+      std::move(promisePair2.second).wait();
+      auto response = std::move(responseFuture).get();
+      ASSERT_EQ(response->headers()->getStatusCode(), http::kHttpOk);
+      ASSERT_EQ(bodyAsString(*response, memoryPool.get()), "Timedout");
+      ASSERT_EQ(request1->requestStatus, kStatusValid);
+    } catch (...) {
+      // The second request shall not timeout at http level. Because time return
+      // for response2 should be 00:03.010 - 00:00.100 = 2910ms < 2950ms
```

Comment:
That's cutting it pretty close, have you run this test a few times to make sure it's passes consistently?

What if the host you're running on is under heavy load (e.g. you're running tests on multiple threads)?

Reply:
Yeah that's a valid concern. I thought about using future and stuff to make the concern go away but that will defy the purpose of the test that is to 1. time out on a time period and 2. more readable so it could be understood more easily on different timeout behavior.

Let me make it more loose.